Fix Windows service installation

coonce opened the pull request:
Some batch files variables weren't expanded at the right time.
'setlocal enabledelayedexpansion' seems to fix the problem so that %start_erl% is set to the proper place within the IF block.

bitwalker commented:
@coonce Thanks! I'm going to be merging a bunch of windows support work done by @OnorioCatenacci    as soon as he gets an updated PR in, and I'd like to have you merge your changes with him once they are in master. I'll update this thread once that's set.

OnorioCatenacci commented:
This PR is a good step but it's not enough.  Oddly enough it's actually sort of orthogonal to the changes I made.  But the problem is that this PR doesn't address is setting the root_dir variable correctly. The fix is certainly valid; it's just not extensive enough to address all of the issues. Try your fix with Erlang installed in c:\program files\erl<ver>  and I think you'll see what I mean.

coonce commented:
Hopefully I tracked down all the instances where quotes were needed (or not). This is working for me with a space in my project name - let me know if you have any luck.

OnorioCatenacci commented:
Hi @coonce,

Please make sure you pull the changes I made which Paul merged because
they're in basically the same parts of the bat files.  I just want to
insure that we don't tromp on each other's changes before we start
reviewing code.

--
Onorio


On Fri, Jan 22, 2016 at 9:03 PM, coonce <notifications@github.com> wrote:

> Hopefully I tracked down all the instances where quotes were needed (or
> not). This is working for me with a space in my project name - let me know
> if you have any luck.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bitwalker/exrm/pull/281#issuecomment-174108848>.
>



-- 
Onorio Catenacci

http://onor.io
http://www.google.com/+OnorioCatenacci

coonce commented:
I found a few more instances of paths that needed to be in quotes, so I've updated this PR. I don't really have a good way to test these changes other than running it manually. So far It Works on My Machine™. :)

OnorioCatenacci commented:
Actually, none of the paths should need to be double quoted now.  I
modified the whole thing to use short (8.3) names everywhere.  I did this
exactly because it's easy to miss quotes and it's also easy to get confused
about where they need to be escaped and where they don't.  In fact adding
double quotes around the new paths may actually cause problems.

I do have a clean test environment for Windows so I'll see if I can grab
the changes you made in this PR and see if it impacts anything either way.
Even if the addition of quotes doesn't have a negative technical impact,
they seem superfluous now and I think they're only making the code a bit
less clear.

I'll post the results of my test when I can.

--
Onorio


On Sun, Jan 24, 2016 at 2:08 PM, coonce <notifications@github.com> wrote:

> I found a few more instances of paths that needed to be in quotes, so I've
> updated this PR. I don't really have a good way to test these changes other
> than running it manually. So far It Works on My Machine™. :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bitwalker/exrm/pull/281#issuecomment-174331774>.
>



-- 
Onorio Catenacci

http://onor.io
http://www.google.com/+OnorioCatenacci

coonce commented:
I agree - the fewer quotes, the better. I updated this again to just include the minimum changes that were required to make this work on my machine.

OnorioCatenacci commented:
Ok, just tried this on a clean Win10 VM.  The install works fine but when I
attempt to start the service I get a "process terminated unexpectedly".  To
be honest, this was exactly the problem I ran into when I was fixing this.
I mean I got the install to work but it kept giving me an error when I
tried to start the process as well.  My changes that Paul pulled in address
this issue.  Take a look at line 187 and 190 of boot.bat to see the fixes I
had to make to address this.

--
Onorio


On Mon, Jan 25, 2016 at 11:17 AM, coonce <notifications@github.com> wrote:

> I agree - the fewer quotes, the better. I updated this again to just
> include the minimum changes that were required to make this work on my
> machine.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bitwalker/exrm/pull/281#issuecomment-174560193>.
>



-- 
Onorio Catenacci

http://onor.io
http://www.google.com/+OnorioCatenacci

coonce commented:
Did you pull in the version with `setlocal enabledelayedexpansion` in the `:install` section? The installation seems to work without it, but it doesn't actually set the arguments properly. Then when you try to start the service, you get the error.
The `%svc_machine%` isn't set when `%erlsrv%` tries to use it to install the service. 
Here's some info on delayed expansion:
[http://superuser.com/a/78509](http://superuser.com/a/78509)

OnorioCatenacci commented:
I pulled from your master branch.  I see that setlocal . . . in the tip of
the master branch.  I'll double check the generated batch file as well.

On Mon, Jan 25, 2016 at 12:19 PM, coonce <notifications@github.com> wrote:

> Did you pull in the version with setlocal enabledelayedexpansion in the
> :install section? The installation seems to work without it, but it
> doesn't actually set the arguments properly. Then when you try to start the
> service, you get the error.
> The %svc_machine% isn't set when %erlsrv% tries to use it to install the
> service.
> Here's some info on delayed expansion:
> http://superuser.com/a/78509
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bitwalker/exrm/pull/281#issuecomment-174591879>.
>



-- 
Onorio Catenacci

http://onor.io
http://www.google.com/+OnorioCatenacci

OnorioCatenacci commented:
Hmm.  May have to clean it all up and repull your code.  The generated
batch file doesn't have that setlocal stuff.

You know though, honestly, I know the master branch code works because I
also tested that on my clean VM image before I submitted the pull request.
I'm not sure what we're gaining by pulling in your changes too since you
and I have both basically addressed the same issue in two different ways.
Is there some strong reason to add the setlocal stuff?  I mean are you
aware of some flaw with my approach or some reason to prefer the setlocal?
If there's something lacking in my approach, please let me know; I'm always
happy to learn.   Otherwise, there are other things still to be resolved on
Windows with Exrm so I'd say we should just let this go and start working
on some of those other problems.


On Mon, Jan 25, 2016 at 12:23 PM, Onorio Catenacci <Catenacci@ieee.org>
wrote:

> I pulled from your master branch.  I see that setlocal . . . in the tip of
> the master branch.  I'll double check the generated batch file as well.
>
> On Mon, Jan 25, 2016 at 12:19 PM, coonce <notifications@github.com> wrote:
>
>> Did you pull in the version with setlocal enabledelayedexpansion in the
>> :install section? The installation seems to work without it, but it
>> doesn't actually set the arguments properly. Then when you try to start the
>> service, you get the error.
>> The %svc_machine% isn't set when %erlsrv% tries to use it to install the
>> service.
>> Here's some info on delayed expansion:
>> http://superuser.com/a/78509
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/bitwalker/exrm/pull/281#issuecomment-174591879>.
>>
>
>
>
> --
> Onorio Catenacci
>
> http://onor.io
> http://www.google.com/+OnorioCatenacci
>
>


-- 
Onorio Catenacci

http://onor.io
http://www.google.com/+OnorioCatenacci

coonce commented:
I was thinking this still needed some attention since it wasn't working on my machine. Sorry for all the noise!

One last question: Should you use `%~dps0` rather than `%dp0` in both `boot.bat` and `boot_shim.bat`? I think `%dps0` would use the 8.3 filename instead.

Thanks for your work on this. Feel free to close the PR.

c

OnorioCatenacci commented:
Maybe we should do the opposite of what I did @coonce.  Pull the latest
version from the tip of the master branch and insure that it works on your
machine.  The more folks with Windows boxes give this a test the more
likely we are to catch anything that I may have missed.

I pretty much used %~dps0 pretty much everywhere so I didn't miss
converting to the 8.3 name of the directory anywhere.  I believe it's
idempotent; I mean to say that I think if the value is long, it gets 8.3
but if it's already 8.3 then it doesn't do anything.

--
Onorio


On Mon, Jan 25, 2016 at 3:23 PM, coonce <notifications@github.com> wrote:

> I was thinking this still needed some attention since it wasn't working on
> my machine. Sorry for all the noise!
>
> One last question: Should you use %~dps0 rather than %dp0 in both boot.bat
> and boot_shim.bat? I think %dps0 would use the 8.3 filename instead.
>
> Thanks for your work on this. Feel free to close the PR.
>
> c
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bitwalker/exrm/pull/281#issuecomment-174650474>.
>



-- 
Onorio Catenacci

http://onor.io
http://www.google.com/+OnorioCatenacci

coonce commented:
I pulled the latest master and some of the commands are still failing for me. I think the culprit is `boot_shim.bat` line 16. The `release_root_dir` variable is being set with the long path name.

Setting the variable with the 8.3 path name seems to fix it:
`set release_root_dir=%%~sfA`

OnorioCatenacci commented:
What are the errors you're getting?  I'd like to add them to my tests.

Would you rather that I make that change and submit another pull request or
would you care to change your code and alter your PR?

--
Onorio






On Mon, Jan 25, 2016 at 4:10 PM, coonce <notifications@github.com> wrote:

> I pulled the latest master and some of the commands are still failing for
> me. I think the culprit is boot_shim.bat line 16. The release_root_dir
> variable is being set with the long path name.
>
> Setting the variable with the 8.3 path name seems to fix it:
> set release_root_dir=%%~sfA
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bitwalker/exrm/pull/281#issuecomment-174664064>.
>



-- 
Onorio Catenacci

http://onor.io
http://www.google.com/+OnorioCatenacci

coonce commented:
Here's what I get with `ELIXIR_CLI_ECHO=y`:
```bash
>rel\spacetest\bin\spacetest install
Administrative permissions required. Detecting permissions...

if 0 == 0 (set is_admin="true" )  else (set is_admin="false" )
The system cannot find the file specified.
Error occurred while processing: e:\dvlp\phoenix\space.
The system cannot find the path specified.

if "true" == "true" (call "e:\dvlp\phoenix\space dir\spacetest\rel\spacetest\releases\0.0.1\spacetest.bat" install )  else ()
The system cannot find the file specified.
Error occurred while processing: e:\dvlp\phoenix\space.
The system cannot find the path specified.

set erts_dir=e:\dvlp\phoenix\SPACED~1\SPACET~1\rel\SPACET~1\erts-7.1

set rootdir=e:\dvlp\phoenix\SPACED~1\SPACET~1\rel\SPACET~1

set sys_config=e:\dvlp\phoenix\space dir\spacetest\rel\spacetest\releases\0.0.1\sys.config
e:\dvlp\phoenix\SPACED~1\SPACET~1\rel\SPACET~1\erts-7.1\bin\erlsrv.exe: Service spacetest_0.0.1 added to system.
```

It looks like it successfully installs. But the `ping` command doesn't work at all:
```bash
>rel\spacetest\bin\spacetest ping
Administrative permissions required. Detecting permissions...

if 0 == 0 (set is_admin="true" )  else (set is_admin="false" )
The system cannot find the file specified.
Error occurred while processing: e:\dvlp\phoenix\space.
The system cannot find the path specified.

if "true" == "true" (call "e:\dvlp\phoenix\space dir\spacetest\rel\spacetest\releases\0.0.1\spacetest.bat" ping )  else ()
The system cannot find the file specified.
Error occurred while processing: e:\dvlp\phoenix\space.
The system cannot find the path specified.

set erts_dir=e:\dvlp\phoenix\SPACED~1\SPACET~1\rel\SPACET~1\erts-7.1

set rootdir=e:\dvlp\phoenix\SPACED~1\SPACET~1\rel\SPACET~1

set sys_config=e:\dvlp\phoenix\space dir\spacetest\rel\spacetest\releases\0.0.1\sys.config
escript: Failed to open file: e:\dvlp\phoenix\space
```

I don't mind updating the PR. I'll see if I can clean up the commits a bit...

coonce committed:
Update boot_shim.bat to use 8.3 path names.

This commit also fixes the ping command by removing the empty quotes
from the conform_opts variable in boot.bat.

OnorioCatenacci commented:
Yeah the ping command doesn't work for me either.  I was more focused on
getting the install/uninstall, start/stop stuff working.  Ping and there's
another one that doesn't work yet either.  That's partly what I meant when
I was saying "there are other things still to be resolved on Windows with
Exrm"  Ping doesn't work and there's another one that doesn't work either
although it escapes me right now.

Either way, I think you're right about getting the 8.3 version of the
directory names in boot_shim.bat.

On Mon, Jan 25, 2016 at 4:38 PM, coonce <notifications@github.com> wrote:

> Here's what I get with ELIXIR_CLI_ECHO=y:
>
> >rel\spacetest\bin\spacetest install
> Administrative permissions required. Detecting permissions...
> if 0 == 0 (set is_admin="true" )  else (set is_admin="false" )
> The system cannot find the file specified.
> Error occurred while processing: e:\dvlp\phoenix\space.
> The system cannot find the path specified.
> if "true" == "true" (call "e:\dvlp\phoenix\space dir\spacetest\rel\spacetest\releases\0.0.1\spacetest.bat" install )  else ()
> The system cannot find the file specified.
> Error occurred while processing: e:\dvlp\phoenix\space.
> The system cannot find the path specified.
> set erts_dir=e:\dvlp\phoenix\SPACED~1\SPACET~1\rel\SPACET~1\erts-7.1
> set rootdir=e:\dvlp\phoenix\SPACED~1\SPACET~1\rel\SPACET~1
> set sys_config=e:\dvlp\phoenix\space dir\spacetest\rel\spacetest\releases\0.0.1\sys.config
> e:\dvlp\phoenix\SPACED~1\SPACET~1\rel\SPACET~1\erts-7.1\bin\erlsrv.exe: Service spacetest_0.0.1 added to system.
>
> It looks like it successfully installs. But the ping command doesn't work
> at all:
>
> >rel\spacetest\bin\spacetest ping
> Administrative permissions required. Detecting permissions...
> if 0 == 0 (set is_admin="true" )  else (set is_admin="false" )
> The system cannot find the file specified.
> Error occurred while processing: e:\dvlp\phoenix\space.
> The system cannot find the path specified.
> if "true" == "true" (call "e:\dvlp\phoenix\space dir\spacetest\rel\spacetest\releases\0.0.1\spacetest.bat" ping )  else ()
> The system cannot find the file specified.
> Error occurred while processing: e:\dvlp\phoenix\space.
> The system cannot find the path specified.
> set erts_dir=e:\dvlp\phoenix\SPACED~1\SPACET~1\rel\SPACET~1\erts-7.1
> set rootdir=e:\dvlp\phoenix\SPACED~1\SPACET~1\rel\SPACET~1
> set sys_config=e:\dvlp\phoenix\space dir\spacetest\rel\spacetest\releases\0.0.1\sys.config
> escript: Failed to open file: e:\dvlp\phoenix\space
>
> I don't mind updating the PR. I'll see if I can clean up the commits a
> bit...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bitwalker/exrm/pull/281#issuecomment-174676183>.
>



-- 
Onorio Catenacci

http://onor.io
http://www.google.com/+OnorioCatenacci

coonce commented:
While messing with the arguments for the service installation, I accidentally got the ping command working. :) The problem was `conform_opts` was set to `""` which was messing with the `-args` in the `:install` section. Should I include this fix in this PR?

OnorioCatenacci commented:
Yes please.  When you get the new PR ready let me know and I'll test on my
Windows 10 VM

On Mon, Jan 25, 2016, 5:56 PM coonce <notifications@github.com> wrote:

> While messing with the arguments for the service installation, I
> accidentally got the ping command working. :) The problem was conform_opts
> was set to "" which was messing with the -args in the :install section.
> Should I include this fix in this PR?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bitwalker/exrm/pull/281#issuecomment-174713446>.
>

coonce commented:
Ok, this should be ready for testing.

OnorioCatenacci commented:
Sorry it took me a little while to get back Coonce.  For the last couple of
weeks or so it's been hard for me to find spare time except at lunch.

It all worked (including the ping!)  Really nice to have that working now
too!

Paul--if you want to pull in Coonce's PR that'll fix the Ping issue so
that'll be one less Windows issue to struggle with.

--
Onorio


On Mon, Jan 25, 2016 at 6:54 PM, coonce <notifications@github.com> wrote:

> Ok, this should be ready for testing.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bitwalker/exrm/pull/281#issuecomment-174735046>.
>



-- 
Onorio Catenacci

http://onor.io
http://www.google.com/+OnorioCatenacci

OnorioCatenacci commented:
By the way, we're probably going to need some sort of readme for Exrm on
Windows.  Whomever tries it may have to grab MSVCR120.dll (it depends on
whether or not they're trying on a box that has Visual Studio already
installed because MSVCR120.dll is the Visual C++ Runtime which comes with
VS 2013).  I keep seeing the issue on my Win 10 VM because, as I say,  I
intentionally keep it really bare bones so I don't have to worry about some
file being present on the machine other than those we package up from Exrm.



On Tue, Jan 26, 2016 at 1:00 PM, Onorio Catenacci <Catenacci@ieee.org>
wrote:

> Sorry it took me a little while to get back Coonce.  For the last couple
> of weeks or so it's been hard for me to find spare time except at lunch.
>
> It all worked (including the ping!)  Really nice to have that working now
> too!
>
> Paul--if you want to pull in Coonce's PR that'll fix the Ping issue so
> that'll be one less Windows issue to struggle with.
>
> --
> Onorio
>
>
> On Mon, Jan 25, 2016 at 6:54 PM, coonce <notifications@github.com> wrote:
>
>> Ok, this should be ready for testing.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/bitwalker/exrm/pull/281#issuecomment-174735046>.
>>
>
>
>
> --
> Onorio Catenacci
>
> http://onor.io
> http://www.google.com/+OnorioCatenacci
>
>


-- 
Onorio Catenacci

http://onor.io
http://www.google.com/+OnorioCatenacci

jerel commented:
So what's the status on this PR?

I'm getting an error when trying to `start` a Windows 7 service:
```
app_0.0.1: Could not start erlang service "app_0.0.1"with commandline 
["C:\Users\user\Desktop\app\rel\app\releases\start_erl.data" 
-service_event ErlSrv_app_0.0.1 -nohup -name app@127.0.0.1 ].
```
along with 
```
app_0.0.1: Could not start erlang machine
```
and it sounds quite similar to what you've been working with here.

jerel commented:
Edit: I patched my `boot.bat` and `boot_shim.bat` with these changes and the service starts successfully!
